common/src/main/java/com/wynntils/handlers/inventory/InventoryHandler.java

```diff
@@ -0,0 +1,449 @@
+/*
+ * Copyright © Wynntils 2021-2024.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.handlers.inventory;
+
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.core.components.Handler;
+import com.wynntils.core.components.Managers;
+import com.wynntils.handlers.inventory.event.InventoryInteractionEvent;
+import com.wynntils.mc.event.ContainerClickEvent;
+import com.wynntils.mc.event.ContainerSetContentEvent;
+import com.wynntils.mc.event.ContainerSetSlotEvent;
+import com.wynntils.models.worlds.event.WorldStateEvent;
+import com.wynntils.utils.mc.McUtils;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import java.util.LinkedList;
+import java.util.List;
+import net.minecraft.network.protocol.game.ServerboundContainerClickPacket;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.inventory.ClickType;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.neoforged.bus.api.EventPriority;
+import net.neoforged.bus.api.SubscribeEvent;
+
+public class InventoryHandler extends Handler {
+    private RunningInteraction running;
+    private final List<PendingInteractionEntry> pending = new LinkedList<>();
+
+    @SubscribeEvent(priority = EventPriority.LOW)
+    public void onSlotClick(ContainerClickEvent event) {
+        if (running != null && running.handle(event)) return;
+
+        AbstractContainerMenu menu = event.getContainerMenu();
+        ItemStack heldStack = menu.getCarried();
+        int slotNum = event.getSlotNum();
+        switch (event.getClickType()) {
+            case PICKUP -> {
+                if (slotNum < 0) { // Slot -999 -> tossed an item outside the GUI
+                    if (!heldStack.isEmpty()) {
+                        if (event.getMouseButton() == 0 || heldStack.getCount() == 1) {
+                            forceSyncLater(menu);
+                        }
+                        expect(new PendingInteraction.ThrowFromHeld(menu, heldStack.copy()));
+                    }
+                } else {
+                    ItemStack slotStack = menu.getSlot(slotNum).getItem();
+                    if (heldStack.isEmpty()) {
+                        if (!slotStack.isEmpty()) {
+                            expect(new PendingInteraction.PickUp(menu, slotNum, slotStack.copy()));
+                        }
+                    } else if (slotStack.isEmpty()) {
+                        if (event.getMouseButton() == 0 || heldStack.getCount() == 1) {
+                            forceSyncLater(menu);
+                        }
+                        expect(new PendingInteraction.Place(menu, slotNum, heldStack.copy()));
+                    } else if (areItemsSimilar(heldStack, slotStack)) {
```

Comment:
I am not sure I understand this `areItemsSimilar` logic. Are you assuming that it is the same item stack if the name and count matches? Or do we need to take some special action if we are dealing with two different stacks of the same kind of item?

Reply:
So what happens if you pick up one stack with e.g. materials and put it down on top of another stack with the same material?

Reply:
So how does this differ from if you were to pick up one stack with materials, and put it down on another stack with another material? Basically, what I'm asking is what would break if you were to remove this `if (areItemsSimilar(heldStack, slotStack))` check.